### PR TITLE
feat: add support for `--no-extra` flag and setting

### LINKED ISSUE #9333: Propose `--no-extra` Flag
The new conflicting depencies handling introduced in 0.5.3 is great.  Able to solve several project challenges with it.

To enhance the feature I'd propose adding a `--no-extra` flag allowing a specific extra group to be excluded when used with `--all-extras`.

With the config below running `uv sync --all-extras`  results in an error.

```
Resolved 3 packages in 1ms
error: extra `extra1`, extra `extra2` are incompatible with the declared conflicts: {`uv-conflicts[extra1]`, `uv-conflicts[extra2]`}
```
I'd propose one option for handling this to allow the user to specify which of the conflicting extras to exclude.  The option for this syntax could be helpful when the project has many optional extras where specifying each individual could become a long chain.

`uv sync --all-extras --no-extra extra2`

Similar functionality already exists for dependency groups with `uv sync --all-groups --no-groups <NO_GROUP>` and this could behave in a similar manner for extras as well.

```toml
[project]
name = "uv-conflicts"
version = "0.1.0"
description = "Add your description here"
readme = "README.md"
requires-python = ">=3.12"
dependencies = []

[project.optional-dependencies]
extra1 = ["numpy==2.1.2"]
extra2 = ["numpy==2.0.0"]

[tool.uv]
conflicts = [
    [
      { extra = "extra1" },
      { extra = "extra2" },
    ],
]
```

### PR DESCRIPTION
**Pull request by alan910127**
## Summary

Resolves #9333  

This pull request introduces support for the `--no-extra` command-line flag and the corresponding `no-extra` UV setting.  

### Behavior
- When `--all-extras` is supplied, the specified extras in `--no-extra` will be excluded from the installation.  
- If `--all-extras` is not supplied, `--no-extra` has no effect and is safely ignored.  

## Test Plan

Since `ExtrasSpecification::from_args` and `ExtrasSpecification::extra_names` are the most important parts in the implementation, I added the following tests in the `uv-configuration/src/extras.rs` module:

- **`test_no_extra_full`**: Verifies behavior when `no_extra` includes the entire list of extras.  
- **`test_no_extra_partial`**: Tests partial exclusion, ensuring only specified extras are excluded.  
- **`test_no_extra_empty`**: Confirms that no extras are excluded if `no_extra` is empty.  
- **`test_no_extra_excessive`**: Ensures the implementation ignores `no_extra` values that don't match any available extras.  
- **`test_no_extra_without_all_extras`**: Validates that `no_extra` has no effect when `--all-extras` is not supplied.  
- **`test_no_extra_without_package_extras`**: Confirms correct behavior when no extras are available in the package.  
- **`test_no_extra_duplicates`**: Verifies that duplicate entries in `pkg_extras` or `no_extra` do not cause errors.

### CODE REVIEW

**Comment by alan910127** [comment]
Hello, uv maintainers, I managed to get the CI to pass, except for the `cargo test | macos` job. The error message indicates it failed to fetch something due to a "too many requests" error. Unfortunately, I wasn’t able to resolve this issue.

**Comment by charliermarsh** [comment]
This looks good, but I think `--extra foo --no-extra foo` should omit `foo`.

File: crates/uv-configuration/src/extras.rs [hunk 1]
```diff
@@ -6,12 +8,15 @@ pub enum ExtrasSpecification {
...
     /// Determine the extras specification to use based on the command-line arguments.
-    pub fn from_args(all_extras: bool, extra: Vec<ExtraName>) -> Self {
-        if all_extras {
+    pub fn from_args(all_extras: bool, no_extra: Vec<ExtraName>, extra: Vec<ExtraName>) -> Self {
+        if all_extras && !no_extra.is_empty() {
+            ExtrasSpecification::Exclude(HashSet::from_iter(no_extra))
+        } else if all_extras {
             ExtrasSpecification::All
         } else if extra.is_empty() {
             ExtrasSpecification::None
```
Comment (charliermarsh) [inline crates/uv-configuration/src/extras.rs:22]: Maybe we filter `extra` in the `else` branch here, to remove anything in `no_extra`?

**Review by charliermarsh — APPROVED** [review]
Excellent PR, thank you!

**Pull request merged**